proxy/src/fs_watch.rs

```diff
@@ -0,0 +1,701 @@
+where
+    I: IntoIterator<Item=P>,
+    P: AsRef<Path>,
+{
+    use ::stream;
+
+    let paths: Vec<PathBuf> = paths.into_iter()
+        .map(|p| p.as_ref().to_path_buf())
+        .collect();
+    let polls = Box::new(stream_changes_polling(paths.clone(), interval));
+    match inotify::WatchStream::new(paths) {
+        Ok(watch) => {
+            let stream = inotify::FallbackStream {
+                watch,
+                polls,
+            };
+            stream::Either::A(stream)
+        },
+        Err(e) => {
+            // If initializing the `Inotify` instance failed, it probably won't
+            // succeed in the future (it's likely that inotify unsupported on
+            // this OS).
+            warn!("inotify init error: {}, falling back to polling", e);
+            stream::Either::B(polls)
+        },
+    }
+}
+
+#[derive(Clone, Debug)]
+struct PathAndHash {
+    /// The path to the file.
+    path: PathBuf,
+
+    /// The last SHA-384 digest of the file, if we have previously hashed it.
+    last_hash: RefCell<Option<Digest>>,
+}
+
+impl PathAndHash {
+    fn new<P: AsRef<Path>>(path: P) -> Self {
+        Self {
+            path: path.as_ref().to_path_buf(),
+            last_hash: RefCell::new(None),
+        }
+    }
+
+    fn has_changed(&self) -> io::Result<bool> {
+        let contents = fs::read(&self.path)?;
+        let hash = Some(digest::digest(&digest::SHA256, &contents[..]));
+        let changed = self.last_hash
+            .borrow().as_ref()
+            .map(Digest::as_ref) != hash.as_ref().map(Digest::as_ref);
+        if changed {
+            self.last_hash.replace(hash);
+        }
+        Ok(changed)
+    }
+}
+
+#[cfg(target_os = "linux")]
+pub mod inotify {
+    use std::{
+        io,
+        path::PathBuf,
+    };
+    use inotify::{
+        Inotify,
+        Event,
+        EventMask,
+        EventStream,
+        WatchMask,
+    };
+    use futures::{Async, Poll, Stream};
+
+    pub struct WatchStream {
+        inotify: Inotify,
+        stream: EventStream,
+        paths: Vec<PathBuf>,
+    }
+
+    pub struct FallbackStream {
+        pub watch: WatchStream,
+        pub polls: Box<Stream<Item = (), Error = ()> + Send>,
+    }
+
+    impl WatchStream {
+        pub fn new(paths: Vec<PathBuf>) -> Result<Self, io::Error> {
+            let mut inotify = Inotify::init()?;
+            let stream = inotify.event_stream();
+
+            let mut watch_stream = WatchStream {
+                inotify,
+                stream,
+                paths,
+            };
+
+            watch_stream.add_paths()?;
+
+            Ok(watch_stream)
+        }
+
+
+        fn add_paths(&mut self) -> Result<(), io::Error> {
+            let mask
+                = WatchMask::CREATE
+                | WatchMask::MODIFY
+                | WatchMask::DELETE
+                | WatchMask::DELETE_SELF
+                | WatchMask::MOVE
+                | WatchMask::MOVE_SELF
```

Comment:
Shouldn't we include `IN_ONLYDIR` too? Otherwise, this looks right to me, though it's not clear to me whether or not we want to use `IN_EXCL_UNLINK`. Something to figure out later.

Reply:
> Shouldn't we include `IN_ONLYDIR` too? 

When a file exists when starting the watch, we _do_ watch the actual file, so I didn't add that mask flag

Reply:
> When a file exists when starting the watch, we do watch the actual file, so I didn't add that mask flag

Ah, yeah. Why do we do that? It seems like it would always be correct to watch the parent directory regardless of whether the file initially exists. Generally unnecessary special cases are bad and "file existed at the beginning" seems like a special case.

Reply:
We always watch the files we care about directly, if they exist. We only watch their parentl directories if they don't currently exist. This is so that we see fewer events on unrelated files, if a file we care about is in a directory that also contains files we _don't_ care about watching.

Reply:
Sorry if my earlier comment wasn't clear; by "when starting a watch", I meant when starting a new _inotify_ watch, not when the watch stream is created.

Reply:
> We always watch the files we care about directly, if they exist. We only watch their parentl directories if they don't currently exist. This is so that we see fewer events on unrelated files, if a file we care about is in a directory that also contains files we don't care about watching.

In our case, won't the directory we are watching only contain files that we care about? I feel like this is an optimization for a general case that isn't actually helpful for us. I would prefer to always do it one way. If we find we're getting too many notifications IRL then we can always add the optimization later.

Reply:
I made an attempt to rewrite this to only watch directories, but I realised that if we're given a path to a file in the CWD (which seems like a fairly plausible case at least while testing locally), we will still sometimes need to watch file paths rather than dirs, so I think there will end up being _some_ special-casing regardless.

The rewritten code looks like this: 
```rust
        fn add_paths(&mut self) -> Result<(), io::Error> {
            let dir_mask
                = WatchMask::CREATE
                | WatchMask::MODIFY
                | WatchMask::DELETE
                | WatchMask::DELETE_SELF
                | WatchMask::MOVE
                | WatchMask::MOVE_SELF
                | WatchMask::ONLYDIR
                ;
            let file_mask
                = WatchMask::CREATE
                | WatchMask::MODIFY
                | WatchMask::DELETE
                | WatchMask::MOVE
                ;
            for path in &self.paths {
                let canonical = path
                    .canonicalize()
                    .unwrap_or_else(|e| {
                        trace!("canonicalize({:?}): {:?}", &path, e);
                        path.to_path_buf()
                    });
                if let Some(ref parent) = canonical.parent() {
                    // If the path has a parent directory, watch that.
                    self.inotify.add_watch(&parent, dir_mask)?;
                    trace!("watch {:?} (for {:?})", parent, path);
                } else {
                    // Otherwise, we're watching the path directly.
                    self.inotify.add_watch(&canonical, file_mask)?;
                    trace!("watch {:?} (for {:?})", &canonical, path);
                };

            }
            Ok(())
        }
```

Let me know if you prefer that over the current implementation, or can think of any ways around having to special-case _something_.

Reply:
> I made an attempt to rewrite this to only watch directories, but I realised that if we're given a path to a file in the CWD (which seems like a fairly plausible case at least while testing locally), we will still sometimes need to watch file paths rather than dirs, so I think there will end up being some special-casing regardless.

I think it's fine to punt on this now. I will file an issue about simplifying it. Basically, I think that we should enforce that these env. vars. to be absolute paths so that the issue of having a directory component of "" goes away. Enforcing them to be absolute paths seems like a good idea anyway. But, I don't want to block landing this on that.